python/rmm/rmm/statistics.py

```diff
@@ -0,0 +1,350 @@
+    current RMM resource stack, this is a no-op.
+
+    Warnings
+    --------
+    This modifies the current RMM memory resource. StatisticsResourceAdaptor
+    is pushed onto the current RMM memory resource stack and must remain the
+    topmost resource throughout the statistics gathering.
+    """
+
+    mr = rmm.mr.get_current_device_resource()
+    if not isinstance(mr, rmm.mr.StatisticsResourceAdaptor):
+        rmm.mr.set_current_device_resource(
+            rmm.mr.StatisticsResourceAdaptor(mr)
+        )
+
+
+def get_statistics() -> Optional[Statistics]:
+    """Get the current allocation statistics.
+
+    Return
+    ------
+    If enabled, returns the current tracked statistics.
+    If disabled, returns None.
+    """
+    mr = rmm.mr.get_current_device_resource()
+    if isinstance(mr, rmm.mr.StatisticsResourceAdaptor):
+        return mr.allocation_counts
+    return None
+
+
+def push_statistics() -> Optional[Statistics]:
+    """Push new counters on the current allocation statistics stack.
+
+    This returns the current tracked statistics and pushes a new set
+    of zero counters on the stack of statistics.
+
+    If statistics are disabled (the current memory resource is not an
+    instance of StatisticsResourceAdaptor), this function is a no-op.
+
+    Return
+    ------
+    If enabled, returns the current tracked statistics _before_ the pop.
+    If disabled, returns None.
+    """
+    mr = rmm.mr.get_current_device_resource()
+    if isinstance(mr, rmm.mr.StatisticsResourceAdaptor):
+        return mr.push_counters()
+    return None
+
+
+def pop_statistics() -> Optional[Statistics]:
+    """Pop the counters of the current allocation statistics stack.
+
+    This returns the counters of current tracked statistics and pops
+    them from the stack.
+
+    If statistics are disabled (the current memory resource is not an
+    instance of StatisticsResourceAdaptor), this function is a no-op.
+
+    Return
+    ------
+    If enabled, returns the popped counters.
+    If disabled, returns None.
+    """
+    mr = rmm.mr.get_current_device_resource()
+    if isinstance(mr, rmm.mr.StatisticsResourceAdaptor):
+        return mr.pop_counters()
+    return None
+
+
+@contextmanager
+def statistics():
+    """Context to enable allocation statistics.
+
+    If statistics has been enabled already (the current memory resource is an
+    instance of StatisticsResourceAdaptor), new counters are pushed on the
+    current allocation statistics stack when entering the context and popped
+    again when exiting using `push_statistics()` and `push_statistics()`.
+
+    If statistics has not been enabled, StatisticsResourceAdaptor is set as
+    the current RMM memory resource when entering the context and removed
+    again when exiting.
+
+    Raises
+    ------
+    ValueError
+        If the current RMM memory source was changed while in the context.
+    """
+
+    prior_non_stats_mr = None
+    if push_statistics() is None:
```

Comment:
*nit* is this _just_ using `push_statistics` returns None to mean "statistics tracking needs to be enabled", or is it also necessary that this push runs in the case where we already have a `StatisticsResourceAdaptor`.

If the former, how about changing `enable_statistics` to return the MR before enablement:

i.e.
```
def enable_statistics() -> rmm.mr.DeviceMemoryResource:
    mr = rmm.mr.get_current_device_resource()
    if not isinstance(mr, rmm.mr.StatisticsResourceAdaptor):
         rmm.mr.set_current_device_resource(rmm.mr.StatisticsResourceAdaptor(mr))
    return mr
```

Then this could just be:

```python
original_mr = enable_statistics()
try:
    current_mr = rmm.mr.get_current_device_resource()
    yield
finally:
    if current_mr is not rmm.mr.get_current_device_resource():
        raise ...
    rmm.mr.set_current_device_resource(original_mr)
```

WDYT?

Reply:
> or is it also necessary that this push runs in the case where we already have a `StatisticsResourceAdaptor`.

This:
```
    If statistics have been enabled already (the current memory resource is an
    instance of StatisticsResourceAdaptor), new counters are pushed on the
    current allocation statistics stack when entering the context and popped
    again when exiting using `push_statistics()` and `push_statistics()`.
```

Reply:
The idea is that you can just use `with statistics():` without worrying about the current mr, you get a fresh pair of counters in any case.

Reply:
Aha, thanks.